temporal-serviceclient/src/main/java/io/temporal/serviceclient/WorkflowServiceStubsOptions.java

```diff
@@ -273,6 +292,87 @@ public Builder setSslContext(SslContext sslContext) {
+     * that require additional customization may use {@link #setSslContext(SslContext)} directly.
+     *
+     * @param keyCertChainInputStream - an input stream for an X.509 client certificate chain in PEM
+     *     format.
+     * @param keyPassword - the password of the keyFile, or null if it's not password-protected.
+     * @param keyInputStream - an input stream for a PKCS#8 client private key in PEM format.
+     * @param fingerprints - a list of SHA1 fingerprints in hexadecimal form of the server's root
+     *     certificate.
+     * @throws SSLException - when it was unable to build the context.
+     */
+    public Builder setCloudSslContext(
+        InputStream keyCertChainInputStream,
+        String keyPassword,
+        InputStream keyInputStream,
+        String... fingerprints)
+        throws SSLException {
+      // TODO consider using default fingerprints from the root CA if user didn't pass any.
+      this.sslContext =
+          SslContextBuilder.forClient()
+              .trustManager(new FingerprintTrustManagerFactory(fingerprints))
+              .keyManager(keyCertChainInputStream, keyInputStream, keyPassword)
+              .applicationProtocolConfig(DEFAULT_APPLICATION_PROTOCOL_CONFIG)
+              .build();
+      return this;
+    }
+
+    /**
+     * Convenience method that overloads {@link #setCloudSslContext(InputStream, String,
+     * InputStream, String...)} and uses no key password.
+     *
+     * @param keyCertChainInputStream - an input stream for an X.509 client certificate chain in PEM
+     *     format.
+     * @param keyInputStream - an input stream for a PKCS#8 client private key in PEM format.
+     * @param fingerprints - a list of SHA1 fingerprints in hexadecimal form of the server's root
+     *     certificate.
+     */
+    public Builder setCloudSslContext(
+        InputStream keyCertChainInputStream, InputStream keyInputStream, String... fingerprints)
+        throws IOException {
+      return setCloudSslContext(keyCertChainInputStream, null, keyInputStream, fingerprints);
+    }
+
+    /**
+     * Sets default SSL context parameters that can be used with the temporal cloud service. Users
+     * that require additional customization may use {@link #setSslContext(SslContext)} directly.
+     *
+     * @param keyCertChainFile - an X.509 client certificate chain file in PEM format.
+     * @param keyPassword - the password of the keyFile, or null if it's not password-protected.
+     * @param keyFile - a PKCS#8 client private key file in PEM format.
+     * @param fingerprints - a list of SHA1 fingerprints in hexadecimal form of the server's root
+     *     certificate.
+     */
+    public Builder setCloudSslContext(
+        File keyCertChainFile, String keyPassword, File keyFile, String... fingerprints)
+        throws IOException {
+      // TODO consider using default fingerprints from the root CA if user didn't pass any.
```

Comment:
if we are connecting to a TLS service with a production LetsEncrypt certificate, and the user doesn't pass any fingerprints, are we saying that the connection will then fail? Or does the FingerprintTrustManagerFactory automatically fallback to the machine's trusted cert store?

Reply:
It does not fallback to anything, underlying implementation is pretty dumb and it simply matches strings.